### PR TITLE
Addition of `black-nb` to CI check to avoid inconsistency in indentation

### LINKED ISSUE #3976: Add Jupyter-black to `check/format`
**Description of the issue**
Some indentation in our Jupyter notebooks are inconsistent (See https://github.com/quantumlib/Cirq/issues/3729). Adding [jupyter-black](https://github.com/drillan/jupyter-black) to our CI will catch this issue early.

### PR DESCRIPTION
**Pull request by paniash**
Fixes #3976 and added quotations around variables. See [here](https://unix.stackexchange.com/questions/68694/when-is-double-quoting-necessary).

### CODE REVIEW

**Comment by paniash** [comment]
@googlebot I signed it!

**Comment by vtomole** [comment]
@paniash Thanks for opening this. Why isn't the [Notebook Formating](https://github.com/quantumlib/Cirq/pull/4020/checks?check_run_id=2334573469) check failing?

**Comment by paniash** [comment]
@vtomole Hmmm.... I'll have to check what's going wrong. Sorry about that. I'll do it over the weekend.

**Comment by vtomole** [comment]
@paniash  The right test is failing but probably not for the right reason. The log says:
```
check/nbformat: line 51: [: missing `]'
check/nbformat: line 51: -z: command not found
```

File: check/nbformat [hunk 1]
```diff
@@ -30,21 +30,29 @@ cd "$(git rev-parse --show-toplevel)"
 
 # Install a pinned version of tf-docs
 TF_DOCS_COMMIT=b3dc8a922d8bdc6e998a8ad4f4953359dc6e576a
-TF_DOCS_INSTALLED=$(pip list | grep  -c $TF_DOCS_COMMIT)
+TF_DOCS_INSTALLED="$(pip list | grep -c "$TF_DOCS_COMMIT")"
+BLACK_INSTALLED="$(pip list | grep -c 'black-nb')"
```
Comment (balopat) [inline check/nbformat:34]: I'd rather have this as a dependency added to dev_tools/conf/pip-list-dependencies.txt

**Comment by paniash** [comment]
@vtomole Hi! Thanks for letting me know. I've fixed it and looks good from my side. Let me know if it's alright. :-)

File: check/nbformat [hunk 2]
```diff
@@ -25,26 +25,34 @@ for arg in $@; do
...
-if [ ! -z "${unformatted}" ]; then
+if [ -n "${unformatted}" ] || [ -n "${black_unformatted}" ]; then
     needed_changes=1
     if (( only_print == 0 )); then
-        $FORMAT_CMD docs
+        "$BLACK_FORMAT" docs
+        "$FORMAT_CMD" docs
     else
         echo -e "\033[31mThe following notebooks require formatting\033[0m."
         echo "${unformatted}"
```
Comment (balopat) [inline check/nbformat:58]: We need to add the list of black unformatted notebooks here too.

Comment (vtomole) [inline check/nbformat:58]: Lets echo `${black_unformatted}` as well.

**Review by vtomole — CHANGES_REQUESTED** [review]
I get 

`./check/nbformat: line 55: python3 -m tensorflow_docs.tools.nbfmt --indent=1: command not found
`

When I call `./check/nbformat --apply`. This is at the end, after all the notebooks have been formatted.

File: check/nbformat [hunk 3]
```diff
@@ -25,26 +25,34 @@ for arg in $@; do
...
+if [ "$BLACK_INSTALLED" != "1" ]; then
+    pip install -U black-nb
+fi
+
 FORMAT_CMD="python3 -m tensorflow_docs.tools.nbfmt --indent=1"
+BLACK_FORMAT="black-nb"
 
 # Test the notebooks
-unformatted=$($FORMAT_CMD --test docs 2>&1 | grep "\- docs" || true)
+unformatted="$("$FORMAT_CMD" --test docs 2>&1 | grep "\- docs" || true)"
```
Comment (balopat) [inline check/nbformat:48]: nbfmt should come after black to avoid breakages in our devsite pipeline

File: check/nbformat [hunk 4]
```diff
@@ -25,26 +25,34 @@ for arg in $@; do
...
 if [ "$TF_DOCS_INSTALLED" != "1" ]; then
   pip install -U git+https://github.com/tensorflow/docs@$TF_DOCS_COMMIT
 fi
 
+if [ "$BLACK_INSTALLED" != "1" ]; then
+    pip install -U black-nb
+fi
+
 FORMAT_CMD="python3 -m tensorflow_docs.tools.nbfmt --indent=1"
+BLACK_FORMAT="black-nb"
```
Comment (balopat) [inline check/nbformat:45]: You're introducing a new format command, but you are not taking into account consistent variable naming - e.g. the other one ends with CMD this one ends with FORMAT. Also, this one is scoped to the tool (BLACK) while the original is left generic "FORMAT_CMD" (which made sense before, as it was the only one). It probably should be renamed by the tool as well - e.g. `TFDOCS_FORMAT`.

**Review by balopat — CHANGES_REQUESTED** [review]
I ran this locally and had a bunch of issues: 

```
check/nbformat: line 51: [: missing `]'
check/nbformat: line 51: -z: command not found
reformatted docs/qcvv/xeb_coherent_noise.ipynb
    14 cells reformatted, 2 cells left unchanged, 1 cell failed to reformat.
reformatted docs/circuits.ipynb
    5 cells reformatted, 12 cells left unchanged, 1 cell failed to reformat.
reformatted docs/protocols.ipynb
    4 cells reformatted, 5 cells left unchanged, 1 cell failed to reformat.
reformatted docs/tutorials/quantum_walks.ipynb
    9 cells reformatted, 4 cells left unchanged, 1 cell failed to reformat.
reformatted docs/tutorials/heatmaps.ipynb
    2 cells reformatted, 1 cell left unchanged, 1 cell failed to reformat.
reformatted docs/tutorials/google/floquet.ipynb
    16 cells reformatted, 17 cells left unchanged, 1 cell failed to reformat.
reformatted docs/qudits.ipynb
    2 cells reformatted, 1 cell failed to reformat.
reformatted docs/qcvv/isolated_xeb.ipynb
    10 cells reformatted, 4 cells left unchanged, 1 cell failed to reformat.
error: cannot format docs/custom_gates.ipynb: INTERNAL ERROR: Black produced different code on the second pass of the formatter.
reformatted docs/tutorials/educators/ion_device.ipynb
    3 cells reformatted, 4 cells left unchanged, 1 cell failed to reformat.
reformatted docs/tutorials/qaoa.ipynb
    13 cells reformatted, 1 cell left unchanged, 1 cell failed to reformat.
reformatted docs/google/concepts.ipynb
    2 cells reformatted, 1 cell failed to reformat.
reformatted docs/tutorials/educators/qaoa_ising.ipynb
    13 cells reformatted, 10 cells left unchanged, 1 cell failed to reformat.
reformatted docs/tutorials/hidden_linear_function.ipynb
    8 cells reformatted, 2 cells left unchanged, 1 cell failed to reformat.
reformatted docs/operators_and_observables.ipynb
    2 cells reformatted, 27 cells left unchanged, 1 cell failed to reformat.
reformatted docs/tutorials/google/colab.ipynb
    4 cells reformatted, 1 cell failed to reformat.
reformatted docs/transform.ipynb
    3 cells reformatted, 1 cell failed to reformat.
error: cannot format docs/tutorials/educators/intro.ipynb: INTERNAL ERROR: Black produced different code on the second pass of the formatter.
reformatted docs/qubits.ipynb
    1 cell reformatted, 1 cell left unchanged, 1 cell failed to reformat.
reformatted docs/tutorials/basics.ipynb
    9 cells reformatted, 6 cells left unchanged, 1 cell failed to reformat.
reformatted docs/tutorials/aqt/getting_started.ipynb
    6 cells reformatted, 1 cell left unchanged, 1 cell failed to reformat.
reformatted docs/simulation.ipynb
    12 cells reformatted, 4 cells left unchanged, 1 cell failed to reformat.
reformatted docs/tutorials/google/echoes.ipynb
    7 cells reformatted, 7 cells left unchanged, 1 cell failed to reformat.
reformatted docs/noise.ipynb
    9 cells reformatted, 14 cells left unchanged, 1 cell failed to reformat.
reformatted docs/tutorials/ionq/getting_started.ipynb
    6 cells reformatted, 3 cells left unchanged, 1 cell failed to reformat.
reformatted docs/interop.ipynb
    5 cells reformatted, 1 cell left unchanged, 1 cell failed to reformat.
reformatted docs/tutorials/google/visualizing_calibration_metrics.ipynb
    18 cells reformatted, 1 cell left unchanged, 1 cell failed to reformat.
reformatted docs/tutorials/google/reservations.ipynb
    10 cells reformatted, 1 cell failed to reformat.
reformatted docs/tutorials/shor.ipynb
    11 cells reformatted, 10 cells left unchanged, 1 cell failed to reformat.
reformatted docs/tutorials/pasqal/getting_started.ipynb
    6 cells reformatted, 6 cells left unchanged, 1 cell failed to reformat.
reformatted docs/tutorials/google/start.ipynb
    7 cells reformatted, 2 cells left unchanged, 1 cell failed to reformat.
reformatted docs/gates.ipynb
    2 cells reformatted, 1 cell failed to reformat.
reformatted docs/tutorials/educators/textbook_algorithms.ipynb
    12 cells reformatted, 20 cells left unchanged, 1 cell failed to reformat.
reformatted docs/tutorials/rabi_oscillations.ipynb
    13 cells reformatted, 2 cells left unchanged, 1 cell failed to reformat.
reformatted docs/tutorials/educators/chemistry.ipynb
    18 cells reformatted, 6 cells left unchanged, 1 cell failed to reformat.
reformatted docs/_template.ipynb
    1 cell reformatted, 1 cell failed to reformat.
reformatted docs/qcvv/parallel_xeb.ipynb
    13 cells reformatted, 5 cells left unchanged, 1 cell failed to reformat.
reformatted docs/tutorials/educators/neutral_atom.ipynb
    7 cells reformatted, 14 cells left unchanged, 1 cell failed to reformat.
reformatted docs/tutorials/variational_algorithm.ipynb
    12 cells reformatted, 3 cells left unchanged, 1 cell failed to reformat.
reformatted docs/qcvv/xeb_theory.ipynb
    10 cells reformatted, 2 cells left unchanged, 1 cell failed to reformat.
All done! ✨ 🍰 ✨
38 files reformatted, 2 files failed to reformat.
check/nbformat: line 55: python3 -m tensorflow_docs.tools.nbfmt --indent=1: command not found

```

Thanks @paniash for working on this, though I'm not convinced yet that this is a good direction. 

I think the first big question is: does black-nb and tf-docs.nbfmt collide with each other or not? By collide I mean that do they agree on the same formatting? If they don't agree, then we could never have a well formatted notebook! 

Second main concern is that we need the tf_docs nbformat to be the **last formatter**, otherwise the site publishing will fail - they do run internally this nbfmt to ensure things are well formatted. 

Third concern is that you haven't seem to have tested the user friendliness of the output in these cases:
1. check/nbformat - should report all the unformatted notebooks by path. The script currently does not do that for the black unformatted notebooks.
```
The following notebooks require formatting.

Notebooks are not formatted. Please run 'check/nbformat --apply'
```
2. `check/nbformat --apply` - there are a ton of issues where it fails on cells, is that okay to ignore or not? Does it report a non zero exit status? If it does, the Github Actions job will fail?

File: check/nbformat [hunk 5]
```diff
@@ -25,29 +25,33 @@ for arg in $@; do
...
 # Get the working directory to the repo root.
-cd "$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
-cd "$(git rev-parse --show-toplevel)"
+cd "$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )" || exit
+cd "$(git rev-parse --show-toplevel)" || exit
 
 # Install a pinned version of tf-docs
 TF_DOCS_COMMIT=b3dc8a922d8bdc6e998a8ad4f4953359dc6e576a
-TF_DOCS_INSTALLED=$(pip list | grep  -c $TF_DOCS_COMMIT)
+TF_DOCS_INSTALLED="$(pip list | grep -c "$TF_DOCS_COMMIT")"
```
Comment (balopat) [inline check/nbformat:33]: shouldn't the internal double quotes be escaped? I'm surprised this works! :)

Comment (paniash) [reply check/nbformat:33]: Should it be? AFAIK, this works fine.

Comment (balopat) [reply check/nbformat:33]: I confirm that this works, but the outer quotes are unnecessary. The output of the command will be stored in the variable no matter if it has spaces or not. I guess we can keep the internal quotes, but again, unnecessary, especially that we define the TF_DOCS_COMMIT variable above.

**Comment by vtomole** [comment]
Hey @paniash, Are you planning on opening a new PR with the changes?

**Pull request closed without merging**